#! /bin/sh /usr/share/dpatch/dpatch-run
## 58_fix_format_select_for_direct_plugins.dpatch by Daniel T Chen <crimsun@ubuntu.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix format selection for direct PCM plugins
## DP: http://hg-mirror.alsa-project.org/alsa-lib?cs=4a9d4e365a94;style=raw

@DPATCH@
diff -urNad alsa-lib-1.0.13~/src/pcm/pcm_direct.c alsa-lib-1.0.13/src/pcm/pcm_direct.c
--- alsa-lib-1.0.13~/src/pcm/pcm_direct.c	2007-02-04 19:47:44.000000000 -0500
+++ alsa-lib-1.0.13/src/pcm/pcm_direct.c	2007-02-04 19:48:00.000000000 -0500
@@ -877,22 +877,26 @@
 	}
 	ret = snd_pcm_hw_params_set_format(spcm, hw_params, params->format);
 	if (ret < 0) {
+		static const snd_pcm_format_t dmix_formats[] = {
+			SND_PCM_FORMAT_S32,
+			SND_PCM_FORMAT_S32 ^ SND_PCM_FORMAT_S32_LE ^ SND_PCM_FORMAT_S32_BE,
+			SND_PCM_FORMAT_S16,
+			SND_PCM_FORMAT_S16 ^ SND_PCM_FORMAT_S16_LE ^ SND_PCM_FORMAT_S16_BE,
+			SND_PCM_FORMAT_S24_3LE,
+		};
 		snd_pcm_format_t format;
-		if (dmix->type == SND_PCM_TYPE_DMIX) {
-			switch (params->format) {
-			case SND_PCM_FORMAT_S32_LE:
-			case SND_PCM_FORMAT_S32_BE:
-			case SND_PCM_FORMAT_S16_LE:
-			case SND_PCM_FORMAT_S16_BE:
-			case SND_PCM_FORMAT_S24_3LE:
+		int i;
+
+		for (i = 0; i < sizeof dmix_formats / sizeof dmix_formats[0]; ++i) {
+			format = dmix_formats[i];
+			ret = snd_pcm_hw_params_set_format(spcm, hw_params, format);
+			if (ret >= 0)
 				break;
-			default:
-				SNDERR("invalid format");
-				return -EINVAL;
-			}
 		}
-		format = params->format;
-		ret = snd_pcm_hw_params_set_format(spcm, hw_params, format);
+		if (ret < 0 && dmix->type != SND_PCM_TYPE_DMIX) {
+			/* TODO: try to choose a good format */
+			ret = snd_pcm_hw_params_set_format_first(spcm, hw_params, &format);
+		}
 		if (ret < 0) {
 			SNDERR("requested or auto-format is not available");
 			return ret;
